tests/kernel/ipi_optimize/src/main.c

```diff
@@ -0,0 +1,741 @@
+}
+
+static void busy_thread_entry(void *p1, void *p2, void *p3)
+{
+	int  key;
+	uint32_t id;
+
+	key = arch_irq_lock();
+	id = _current_cpu->id;
+	arch_irq_unlock(key);
+
+	while (1) {
+		atomic_or(&in_loop, BIT(id));
+
+		k_busy_wait(1000);
+	}
+}
+
+static bool wait_until_busy_threads_ready(uint32_t id)
+{
+	uint32_t  all;
+	uint32_t  value;
+	unsigned int i;
+
+	all = ((1 << CONFIG_MP_MAX_NUM_CPUS) - 1) ^ BIT(id);
+	for (i = 0; i < 10; i++) {
+		k_busy_wait(1000);
+
+		value = (uint32_t)atomic_get(&in_loop);
+		if (value == all) {
+			break;
+		}
+	}
+
+	return (i < 10);
+}
+
+/**
+ * Verify that arch_sched_ipi() directs/broadcasts IPIs as expected.
+ */
+ZTEST(ipi, test_arch_sched_ipi)
+{
+	uint32_t  set[CONFIG_MP_MAX_NUM_CPUS];
+	uint32_t  id;
+	int priority;
+	int key;
+	unsigned int i;
+	unsigned int j;
+
+	priority = k_thread_priority_get(k_current_get());
+	atomic_clear(&in_loop);
+
+	for (i = 0; i < NUM_THREADS; i++) {
+		k_thread_create(&thread[i], stack[i], STACK_SIZE,
+				busy_thread_entry,
+				NULL, NULL, NULL,
+				priority - 1, 0, K_NO_WAIT);
+	}
+
+	/* Align to tick boundary to minimize probability of timer ISRs */
+
+	k_sleep(K_TICKS(1));
+
+	key = arch_irq_lock();
+	id = _current_cpu->id;
+	arch_irq_unlock(key);
+
+	/*
+	 * Spin until all busy threads are ready. It is assumed that as this
+	 * thread and the busy threads are cooperative that they will not be
+	 * rescheduled to execute on a different CPU.
+	 */
+
+	zassert_true(wait_until_busy_threads_ready(id),
+		     "1 or more 'busy threads' not ready.\n");
+
+#if CONFIG_ARCH_HAS_DIRECTED_IPIS
+	/*
+	 * Send an IPI to each CPU, one at a time. Verify that only the
+	 * targeted CPU received the IPI.
+	 */
+	for (i = 0; i < CONFIG_MP_MAX_NUM_CPUS; i++) {
+		if (i == id) {
+			continue;
+		}
+
+		clear_ipi_counts();
+		arch_sched_ipi(BIT(i));
+		k_busy_wait(100);
+		get_ipi_counts(set, CONFIG_MP_MAX_NUM_CPUS);
+
+		for (j = 0; j < CONFIG_MP_MAX_NUM_CPUS; j++) {
+			if (i == j) {
+				zassert_true(set[j] == 1,
+					     "Direct-Expected 1, got %u\n",
+					     set[j]);
+			} else {
+				zassert_true(set[j] == 0,
+					     "Direct-Expected 0, got %u\n",
+					     set[j]);
+			}
+		}
+	}
+#else
+
+	/* Broadcast the IPI. All other CPUs ought to receive and process it */
+
+	clear_ipi_counts();
+	arch_sched_ipi(0);
+	k_busy_wait(100);
+	get_ipi_counts(set, CONFIG_MP_MAX_NUM_CPUS);
+
+	for (j = 0; j < CONFIG_MP_MAX_NUM_CPUS; j++) {
+		if (id == j) {
+			zassert_true(set[j] == 0,
+				     "Broadcast-Expected 0, got %u\n",
+				     set[j]);
+		} else {
+			zassert_true(set[j] == 1,
+				     "Broadcast-Expected 1, got %u\n",
+				     set[j]);
+		}
+	}
+#endif
+}
+
+static void pending_thread_entry(void *p1, void *p2, void *p3)
+{
+	k_sem_take(&sem, K_FOREVER);
+}
+
+/**
+ * Verify that waking a thread whose priority is lower than any other
+ * currently executing thread does not result in any IPIs being sent.
+ */
+ZTEST(ipi, test_low_thread_wakes_no_ipis)
```

Comment:
Seems like a *lot* of boilerplate commonality between these tests.  Can't you refactor such that the thread spawning and IPI trace collection is all common, and just takes a config for "what priorities to use" and "what CPUs should get IPIs"?

Reply:
I'll see what I can do. :)